crates/worker/src/partition/snapshots/repository.rs

```diff
@@ -0,0 +1,698 @@
+        }
+
+        let metadata_key = object_store::path::Path::from(format!(
+            "{}/metadata.json",
+            full_snapshot_path.as_str()
+        ));
+        let metadata_json_payload = PutPayload::from(
+            serde_json::to_string_pretty(snapshot).expect("Can always serialize JSON"),
+        );
+
+        let put_result = self
+            .object_store
+            .put(&metadata_key, metadata_json_payload)
+            .await
+            .map_err(|e| PutSnapshotError::from(e, &progress))?;
+        progress.push("/metadata.json".to_owned());
+
+        debug!(
+            etag = put_result.e_tag.unwrap_or_default(),
+            key = ?metadata_key,
+            "Successfully published snapshot metadata",
+        );
+
+        let latest_path = object_store::path::Path::from(format!(
+            "{prefix}{partition_id}/latest.json",
+            prefix = self.prefix,
+            partition_id = snapshot.partition_id,
+        ));
+
+        // By performing a CAS on the latest snapshot pointer, we can ensure strictly monotonic updates.
+        let maybe_stored = self
+            .get_latest_snapshot_metadata_for_update(snapshot, &latest_path)
+            .await
+            .map_err(|e| PutSnapshotError::from(e, &progress))?;
+        if maybe_stored.as_ref().is_some_and(|(latest_stored, _)| {
+            latest_stored.min_applied_lsn >= snapshot.min_applied_lsn
+        }) {
+            let repository_latest_lsn = maybe_stored.expect("is some").0.min_applied_lsn;
+            info!(
+                ?repository_latest_lsn,
+                new_snapshot_lsn = ?snapshot.min_applied_lsn,
+                "The newly uploaded snapshot is no newer than the already-stored latest snapshot, will not update latest pointer"
+            );
+            return Ok(());
+        }
+
+        // Construct the new "latest snapshot" pointer
+        let latest = LatestSnapshot {
+            version: snapshot.version,
+            cluster_name: snapshot.cluster_name.clone(),
+            node_name: snapshot.node_name.clone(),
+            partition_id: snapshot.partition_id,
+            snapshot_id: snapshot.snapshot_id,
+            created_at: snapshot.created_at.clone(),
+            min_applied_lsn: snapshot.min_applied_lsn,
+            path: relative_snapshot_path,
+        };
+        let latest_json = PutPayload::from(
+            serde_json::to_string_pretty(&latest)
+                .map_err(|e| PutSnapshotError::from(e, &progress))?,
+        );
+
+        // The object_store file provider supports create-if-not-exists but not update-version on put
+        let use_conditional_update = !matches!(self.destination.scheme(), "file");
+        let conditions = maybe_stored
+            .map(|(_, version)| PutOptions {
+                mode: match use_conditional_update {
+                    true => PutMode::Update(version),
+                    false => PutMode::Overwrite,
```

Comment:
One way to solve this problem could be to use a file lock if we want this to work.

Reply:
I think the scenario you describe is absolutely possible if you run multiple `restate-server` processes on the same host, AND configure the same directory as the snapshot destination. I am assuming that `file://` destinations may be used for either local-only single node dev deployments, or perhaps in legacy environment multi-node deployments where you don't have an object store but do have a shared network filesystem available. In the latter case, it would be very useful to have the extra file-based locking mechanism. Do you mind if I do this as a follow-up issue?

Reply:
Created https://github.com/restatedev/restate/issues/2384 to track this :-)

Reply:
Sure, we can do this as a follow-up.